Requested index is unavailable due to compaction

saw this several times when running tests

```
2021/08/18 20:37:49.869480 [error] [transport]: session 2[127.0.0.1:61033] read failed with read tcp 127.0.0.1:10002->127.0.0.1:61033: i/o timeout
{"level":"panic","ts":"2021-08-18T20:37:50.169+0800","caller":"raft/raft.go:927","msg":"unexpected error getting unapplied entries (requested index is unavailable due to compaction)","stacktrace":"go.etcd.io/etcd/pkg/logutil.(*zapRaftLogger).Panicf\n\t/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/pkg/logutil/zap_raft.go:101\ngo.etcd.io/etcd/raft.(*raft).Step\n\t/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/raft/raft.go:927\ngo.etcd.io/etcd/raft.(*raft).tickElection\n\t/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/raft/raft.go:675\ngo.etcd.io/etcd/raft.(*RawNode).Tick\n\t/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/raft/rawnode.go:59\ngithub.com/matrixorigin/matrixcube/raftstore.(*peerReplica).handleTick\n\t/Users/lni/src/matrixcube/raftstore/peer_event_loop.go:288\ngithub.com/matrixorigin/matrixcube/raftstore.(*peerReplica).handleEvent\n\t/Users/lni/src/matrixcube/raftstore/peer_event_loop.go:183\ngithub.com/matrixorigin/matrixcube/raftstore.(*store).runPRTask.func1.1\n\t/Users/lni/src/matrixcube/raftstore/store.go:417\nsync.(*Map).Range\n\t/usr/local/go/src/sync/map.go:346\ngithub.com/matrixorigin/matrixcube/raftstore.(*store).runPRTask.func1\n\t/Users/lni/src/matrixcube/raftstore/store.go:415\ngithub.com/matrixorigin/matrixcube/raftstore.(*store).runPRTask\n\t/Users/lni/src/matrixcube/raftstore/store.go:440\ngithub.com/matrixorigin/matrixcube/raftstore.(*store).startRaftWorkers.func1\n\t/Users/lni/src/matrixcube/raftstore/store.go:403\ngithub.com/fagongzi/util/task.(*Runner).RunCancelableTask.func1\n\t/Users/lni/go/pkg/mod/github.com/fagongzi/util@v0.0.0-20210409031311-a10fdf8fbd7a/task/task.go:350"}
panic: unexpected error getting unapplied entries (requested index is unavailable due to compaction)

goroutine 14599 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0x140001b5c30, {0x0, 0x0, 0x0})
	/Users/lni/go/pkg/mod/go.uber.org/zap@v1.10.0/zapcore/entry.go:229 +0x404
go.uber.org/zap.(*SugaredLogger).log(0x14000364d40, 0x4, {0x101c533b1, 0x2f}, {0x14002f083e0, 0x1, 0x1}, {0x0, 0x0, 0x0})
	/Users/lni/go/pkg/mod/go.uber.org/zap@v1.10.0/sugar.go:234 +0x158
go.uber.org/zap.(*SugaredLogger).Panicf(...)
	/Users/lni/go/pkg/mod/go.uber.org/zap@v1.10.0/sugar.go:159
go.etcd.io/etcd/pkg/logutil.(*zapRaftLogger).Panicf(0x14002fe3430, {0x101c533b1, 0x2f}, {0x14002f083e0, 0x1, 0x1})
	/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/pkg/logutil/zap_raft.go:101 +0x64
go.etcd.io/etcd/raft.(*raft).Step(0x140028eb680, {0x0, 0x0, 0xa, 0x0, 0x0, 0x0, {0x0, 0x0, 0x0}, ...})
	/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/raft/raft.go:927 +0xd08
go.etcd.io/etcd/raft.(*raft).tickElection(0x140028eb680)
	/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/raft/raft.go:675 +0xc4
go.etcd.io/etcd/raft.(*RawNode).Tick(...)
	/Users/lni/go/pkg/mod/github.com/deepfabric/etcd@v1.4.15/raft/rawnode.go:59
github.com/matrixorigin/matrixcube/raftstore.(*peerReplica).handleTick(0x140053aef00, {0x140053c6000, 0x400, 0x400})
	/Users/lni/src/matrixcube/raftstore/peer_event_loop.go:288 +0xe8
github.com/matrixorigin/matrixcube/raftstore.(*peerReplica).handleEvent(0x140053aef00)
	/Users/lni/src/matrixcube/raftstore/peer_event_loop.go:183 +0x140
github.com/matrixorigin/matrixcube/raftstore.(*store).runPRTask.func1.1({0x101e90f20, 0x1029cdef0}, {0x10208bd00, 0x140053aef00})
	/Users/lni/src/matrixcube/raftstore/store.go:417 +0x68
sync.(*Map).Range(0x14003047938, 0x14000b4de28)
	/usr/local/go/src/sync/map.go:346 +0x304
github.com/matrixorigin/matrixcube/raftstore.(*store).runPRTask.func1()
	/Users/lni/src/matrixcube/raftstore/store.go:415 +0x84
github.com/matrixorigin/matrixcube/raftstore.(*store).runPRTask(0x14003047860, {0x1020d3918, 0x14004e7fb80}, 0x0, 0x0)
	/Users/lni/src/matrixcube/raftstore/store.go:440 +0x214
github.com/matrixorigin/matrixcube/raftstore.(*store).startRaftWorkers.func1({0x1020d3918, 0x14004e7fb80})
	/Users/lni/src/matrixcube/raftstore/store.go:403 +0x48
github.com/fagongzi/util/task.(*Runner).RunCancelableTask.func1(0x140030c6410, 0x14004f64620, {0x1020d3918, 0x14004e7fb80})
	/Users/lni/go/pkg/mod/github.com/fagongzi/util@v0.0.0-20210409031311-a10fdf8fbd7a/task/task.go:350 +0x88
created by github.com/fagongzi/util/task.(*Runner).RunCancelableTask
	/Users/lni/go/pkg/mod/github.com/fagongzi/util@v0.0.0-20210409031311-a10fdf8fbd7a/task/task.go:345 +0x148
FAIL	github.com/matrixorigin/matrixcube/raftstore	146.040s
FAIL
make: *** [test] Error 1
```

--- a/config/config.go
+++ b/config/config.go
@@ -414,6 +414,9 @@ type StoreHeartbeatDataProcessor interface {
 
 // TestConfig all test config
 type TestConfig struct {
+	// PeerReplicaDelegateWait sleep before add PeerReplicaDelegate to store
+	PeerReplicaDelegateWait time.Duration
+	// Shards test config for shards
 	Shards map[uint64]*TestShardConfig
 }
 

--- a/raftstore/peer_event_raft_ready.go
+++ b/raftstore/peer_event_raft_ready.go
@@ -382,14 +382,14 @@ func (pr *peerReplica) doApplySnapshot(ctx *readyContext, rd *raft.Ready) *apply
 }
 
 func (pr *peerReplica) applyCommittedEntries(rd *raft.Ready) {
-	// make sure the delegate already registered
-	if _, ok := pr.store.delegates.Load(pr.shardID); !ok {
-		return
-	}
-
 	if pr.ps.isApplyingSnapshot() {
 		pr.ps.lastReadyIndex = pr.ps.getTruncatedIndex()
 	} else {
+		// make sure the delegate already registered
+		if _, ok := pr.store.delegates.Load(pr.shardID); !ok {
+			return
+		}
+
 		if testMaxOnceCommitEntryCount > 0 &&
 			testMaxOnceCommitEntryCount < len(rd.CommittedEntries) {
 			rd.CommittedEntries = rd.CommittedEntries[:testMaxOnceCommitEntryCount]

--- a/raftstore/peer_event_raft_ready_test.go
+++ b/raftstore/peer_event_raft_ready_test.go
@@ -1,0 +1,41 @@
+package raftstore
+
+import (
+	"testing"
+	"time"
+
+	"github.com/matrixorigin/matrixcube/config"
+	"github.com/matrixorigin/matrixcube/pb/bhmetapb"
+	"github.com/matrixorigin/matrixcube/util/leaktest"
+)
+
+func TestIssue133(t *testing.T) {
+	// time line
+	// 1. leader send message to a follower
+	// 2. follower create with empty apply state
+	// 3. leader send snapshot to follower
+	// 4. receive the snapshot message
+	// 5. advance raft with 0 applied index
+	// 6. peerReplica delegate create
+
+	defer leaktest.AfterTest(t)()
+	c := NewTestClusterStore(t, WithAppendTestClusterAdjustConfigFunc(func(node int, cfg *config.Config) {
+		cfg.Customize.CustomInitShardsFactory = func() []bhmetapb.Shard {
+			return []bhmetapb.Shard{
+				{Start: []byte("a"), End: []byte("b")},
+				{Start: []byte("b"), End: []byte("c")},
+				{Start: []byte("c"), End: []byte("d")},
+			}
+		}
+
+		if node > 0 {
+			cfg.Test.PeerReplicaDelegateWait = time.Second
+		}
+	}))
+	defer c.Stop()
+
+	c.Start()
+	c.WaitShardByCounts(t, [3]int{3, 3, 3}, time.Second*10)
+	c.StopNode(0)
+	time.Sleep(time.Second)
+}

--- a/raftstore/peer_job.go
+++ b/raftstore/peer_job.go
@@ -50,6 +50,9 @@ func (pr *peerReplica) startRegistrationJob() {
 	}
 
 	err := pr.store.addApplyJob(pr.applyWorker, "doRegistrationJob", func() error {
+		if pr.store.cfg.Test.PeerReplicaDelegateWait > 0 {
+			time.Sleep(pr.store.cfg.Test.PeerReplicaDelegateWait)
+		}
 		return pr.doRegistrationJob(delegate)
 	}, nil)
 

--- a/raftstore/store.go
+++ b/raftstore/store.go
@@ -120,8 +120,9 @@ type store struct {
 	writeHandlers map[uint64]command.WriteCommandFunc
 	localHandlers map[uint64]command.LocalCommandFunc
 
-	stopWG sync.WaitGroup
-	state  uint32
+	stopWG   sync.WaitGroup
+	state    uint32
+	stopOnce sync.Once
 
 	localCB func(*raftcmdpb.RaftResponseHeader, *raftcmdpb.Response)
 	rpcCB   func(*raftcmdpb.RaftResponseHeader, *raftcmdpb.Response)
@@ -196,18 +197,20 @@ func (s *store) Start() {
 func (s *store) Stop() {
 	atomic.StoreUint32(&s.state, 1)
 
-	s.pd.Stop()
-	s.foreachPR(func(pr *peerReplica) bool {
-		s.stopWG.Add(1)
-		pr.stopEventLoop()
-		return true
-	})
-	s.stopWG.Wait()
+	s.stopOnce.Do(func() {
+		s.pd.Stop()
+		s.foreachPR(func(pr *peerReplica) bool {
+			s.stopWG.Add(1)
+			pr.stopEventLoop()
+			return true
+		})
+		s.stopWG.Wait()
 
-	s.snapshotManager.Close()
-	s.runner.Stop()
-	s.trans.Stop()
-	s.rpc.Stop()
+		s.snapshotManager.Close()
+		s.runner.Stop()
+		s.trans.Stop()
+		s.rpc.Stop()
+	})
 }
 
 func (s *store) prStopped() {

--- a/raftstore/store_test.go
+++ b/raftstore/store_test.go
@@ -36,7 +36,7 @@ import (
 
 func TestClusterStartAndStop(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	c := NewTestClusterStore(t, WithTestClusterLogLevel("info"))
+	c := NewTestClusterStore(t)
 	defer c.Stop()
 
 	c.Start()

--- a/raftstore/testutil.go
+++ b/raftstore/testutil.go
@@ -109,7 +109,7 @@ func (opts *testClusterOptions) adjust() {
 		opts.nodes = 3
 	}
 	if opts.logLevel == "" {
-		opts.logLevel = "error"
+		opts.logLevel = "info"
 	}
 }
 

